diff --git a/constructs/NextjsSite.js b/constructs/NextjsSite.js
index 1af9bdaa201f0c52e40ec86defef753d979a9d19..e62c17ae66d224ab39120ba9ab727ef7ebb7a960 100644
--- a/constructs/NextjsSite.js
+++ b/constructs/NextjsSite.js
@@ -37,7 +37,7 @@ export class NextjsSite extends SsrSite {
         };
     }
     createFunctionForRegional() {
-        const { runtime, timeout, memorySize, permissions, environment, cdk } = this.props;
+        const { runtime, timeout, memorySize, permissions, environment, cdk, bind } = this.props;
         const ssrFn = new SsrFunction(this, `ServerFunction`, {
             description: "Server handler for Next.js",
             bundlePath: path.join(this.props.path, ".open-next", "server-function"),
@@ -47,6 +47,7 @@ export class NextjsSite extends SsrSite {
             memorySize,
             permissions,
             environment,
+            bind,
             ...cdk?.server,
         });
         return ssrFn.function;